internal/ethapi/transaction_args.go

```diff
@@ -139,10 +169,31 @@ func (args *TransactionArgs) ToMessage(globalGasCap uint64) types.Message {
 		log.Warn("Caller gas above allowance, capping", "requested", gas, "cap", globalGasCap)
 		gas = globalGasCap
 	}
-	gasPrice := new(big.Int)
+	var gasPrice *big.Int
 	if args.GasPrice != nil {
 		gasPrice = args.GasPrice.ToInt()
 	}
+	var feeCap *big.Int
+	if args.FeeCap != nil {
+		feeCap = args.FeeCap.ToInt()
+	}
+	var tip *big.Int
+	if args.Tip != nil {
+		tip = args.Tip.ToInt()
+	}
+	if gasPrice == nil && feeCap == nil && tip == nil {
+		log.Warn("At least one of gasPrice, feeCap, tip should be set. Defaulting to 0 gas price.")
+		gasPrice = new(big.Int)
+	}
+	if gasPrice != nil && (feeCap != nil || tip != nil) {
+		log.Warn("Only one of gasPrice or feeCap+tip should be set. Ignoring gasPrice, using feeCap+tip")
+		gasPrice = nil
+	}
+	// If baseFee provided, set gasPrice to effectiveGasPrice.
+	if baseFee != nil {
+		gasPrice = math.BigMin(new(big.Int).Add(tip, baseFee), feeCap)
```

Comment:
If I have a baseFee and gasPrice != nil && tip == nil && fee == nil, I'd expect this to ether blow up, or be very incorrect :)

Reply:
Yeah, it's gonna panic https://play.golang.org/p/Xxfr6yqdwNK

Reply:
Bleh, this method is so messy, particularly since it doesn't return an error

Reply:
Ok, pushed a fix

Reply:
I'm not convinced it's ok though. If gasPrice *is* specified, it should be treated as a legacy tx and not overridden.

Reply:
Here's a counter proposal that should clean things up a bit https://gist.github.com/karalabe/91716ea8ae7e7e9fa81ffcee7d68a8b4

Reply:
@karalabe this is to simulate the behaviour in `Transaction.AsMessage`

https://github.com/ethereum/go-ethereum/blob/2cde472650e27634b9e2243d687629dfcffa211a/core/types/transaction.go#L608-L611

I believe this is only used to propogate the `effectiveTip` to the `gasprice` opcode. It may be better to just have this implemented in the `state_transition` logic rather than `AsMessage`?

Reply:
The idea is that it's ok to try and be smart and guess non-specified fields, but if we got invalid combos, I feel it's a bad idea to "fix it" automagically, we should just reject it.

Reply:
That makes sense.